osutil/epoll/epoll.go

```diff
@@ -188,13 +189,66 @@ type Event struct {
 	Readiness Readiness
 }
 
+// Wait blocks and waits for arrival of events on any of the added file descriptors.
+func (e *Epoll) Wait() ([]Event, error) {
+	duration := time.Duration(-1)
+	return e.WaitTimeout(duration)
+}
+
+// WaitTimeout blocks and waits with the given timeout for arrival of events on any of the added file descriptors.
+//
+// A duration value of -1 milliseconds disables timeout.
+//
+// Please refer to epoll_wait(2) and EPOLL_WAIT for details.
+//
+// Warning, using epoll from Golang explicitly is tricky.
+func (e *Epoll) WaitTimeout(duration time.Duration) ([]Event, error) {
+	return e.WaitTimeoutWithContext(nil, duration)
+}
+
+// Wait blocks and waits for arrival of events on any of the added file descriptors, or for the given context to be done.
+func (e *Epoll) WaitWithContext(ctx context.Context) ([]Event, error) {
+	duration := time.Duration(-1)
+	return e.WaitTimeoutWithContext(ctx, duration)
+}
+
+// WaitTimeoutWithContext blocks and waits with the given timeout for arrival of events on any of the added file descriptors, or for the given context to be done.
+//
+// A duration value of -1 milliseconds disables timeout.
+//
+// Please refer to epoll_wait(2) and EPOLL_WAIT for details.
+//
+// Warning, using epoll from Golang explicitly is tricky.
+func (e *Epoll) WaitTimeoutWithContext(ctx context.Context, duration time.Duration) ([]Event, error) {
```

Comment:
I considered this, but I want to pass the timeout into the syscall itself, so it doesn't have to keep retrying the syscall with short timeouts, and so that it doesn't consume a signal after the context times out but before the syscall times out (with some default duration).

It would make the logic to compute the remaining timeout after `EINTR` much easier though.

I also considered creating a `context.WithTimeout` with the give timeout and passing it into the syscall, but this seems more complicated and confusing than just letting the syscall deal with the timeout.

Reply:
I see, thank you. I didn't take into account that you are passing the duration to `unixEpollWait`.

Reply:
Importantly though, there could be a signal lost though if the syscall returns an epoll event after the ~~timeout or~~ context expires. We need to store these in the epoll instance somehow until another `Wait` call picks them up (with timeout of 0 passed to the syscall) or until the epoll instance is closed.

I think there's definitely enough to tweak about epoll to make a separate PR, which can include the epoll changes here as well as the fixes to prevent signal loss.

Reply:
There's no need, it turns out. The only time we want to cancel an `epoll_wait` call syscall is when we're closing the listener, in which case we want to close the epoll instance as well. When closing the epoll instance, all `epoll.Wait` calls return immediately, and we can then check if the listener is closed (or the tomb, for that matter) and return accordingly. Then from the `listener.Close` function, after closing the epoll instance, calling `l.tomb.Wait()` waits for the remaining waiters to terminate, and then can return knowing that the listener is fully closed.

As mentioned previously, I think epoll_wait isn't really compatible with passed-in contexts, as this could cause signal loss that is verry difficult to work around. So I'm going to revert the epoll modifications commit, with the intention of rebasing the PR before merging to delete the epoll PR and its reversion.